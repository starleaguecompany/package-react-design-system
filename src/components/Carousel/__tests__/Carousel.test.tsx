import * as React from 'react'
import { render, screen } from '@testing-library/react'
import { mocked } from 'jest-mock'

import { Carousel } from '../index'
import userEvent from '@testing-library/user-event'

const mockReact = React

jest.mock('react', () => {
  return {
    ...jest.requireActual<typeof mockReact>('react'),
    useRef: jest.fn(),
  }
})

const useMockRef = mocked(React.useRef)

describe('Carousel', () => {
  beforeEach(() => {
    const ref = { current: {} }
    Object.defineProperty(ref, 'current', {
      set(_current) {
        if (_current) {
          jest.spyOn(_current, 'offsetWidth', 'get').mockReturnValue(100)
          jest.spyOn(_current, 'scrollWidth', 'get').mockReturnValue(200)
        }
        this._current = _current
      },
      get() {
        return this._current
      },
    })
    useMockRef.mockReturnValueOnce(ref)
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  afterAll(() => {
    jest.resetAllMocks()
  })

  test('should render correctly with no props', () => {
    render(<Carousel />)

    expect(screen.getByTestId('Carousel')).toMatchSnapshot()
    expect(screen.getByTestId('Carousel')).toHaveClass('container', { exact: true })
  })

  test('should render correctly with slides', () => {
    render(
      <Carousel>
        {Array.from(new Array(3)).map((_, key) => (
          <div key={key}>{key + 1}</div>
        ))}
      </Carousel>
    )

    expect(screen.getByTestId('Carousel')).toMatchSnapshot()
    expect(screen.getByText('1')).toBeInTheDocument()
    expect(screen.getByText('2')).toBeInTheDocument()
    expect(screen.getByText('3')).toBeInTheDocument()
    expect(screen.getByTestId('ArrowLeft')).toBeInTheDocument()
    expect(screen.getByTestId('ArrowRight')).toBeInTheDocument()
  })

  test('should render with title', () => {
    render(<Carousel title="Title"></Carousel>)

    expect(screen.getByTestId('Carousel')).toMatchSnapshot()
    expect(screen.getByText('Title')).toBeInTheDocument()
  })

  test('should render correctly with slides and gradient', () => {
    const { container } = render(
      <Carousel gradient>
        {Array.from(new Array(3)).map((_, key) => (
          <div key={key} style={{ width: '300px' }}>
            {key + 1}
          </div>
        ))}
      </Carousel>
    )

    expect(screen.getByTestId('Carousel')).toMatchSnapshot()
    expect(container.querySelector('.slides')).toHaveClass('gradient')
  })

  test('should move slide', () => {
    const { container } = render(
      <Carousel gradient>
        {Array.from(new Array(4)).map((_, key) => (
          <div key={key} style={{ width: '300px' }}>
            {key + 1}
          </div>
        ))}
      </Carousel>
    )

    expect(screen.getByTestId('Carousel')).toMatchSnapshot()
    //@ts-ignore
    userEvent.click(container.querySelector('.rightArrow'))
    expect(container.querySelector('.slides')).toHaveClass('leftGradient')
    expect(container.querySelector('.rightArrow')).not.toHaveClass('disabled')
  })
})
