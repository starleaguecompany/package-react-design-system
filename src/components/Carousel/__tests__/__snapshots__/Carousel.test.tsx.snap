// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel should move slide 1`] = `
<div
  class="container"
  data-qa="Carousel"
  style="--position: 0px;"
>
  <div
    class="header container direction-horizontal justify-space-between"
    data-qa="Space"
  >
    <div
      class="title"
    />
    <div
      class="arrows container direction-horizontal align-end"
      data-qa="Space"
      style="column-gap: 12px; row-gap: 12px;"
    >
      <span
        class="arrow leftArrow disabled container size-16 asContainer shape-round"
        data-qa="Icon"
      >
        <svg
          data-qa="ArrowLeft"
          fill="none"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m13 4-6 6 6 6"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </span>
      <span
        class="arrow rightArrow container size-16 asContainer shape-round"
        data-qa="Icon"
      >
        <svg
          data-qa="ArrowRight"
          fill="none"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m7 16 6-6-6-6"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    class="slides gradient rightGradient"
  >
    <div
      class="wrapper"
    >
      <div
        style="width: 300px;"
      >
        <div
          style="width: 300px;"
        >
          1
        </div>
      </div>
      <div
        style="width: 300px;"
      >
        <div
          style="width: 300px;"
        >
          2
        </div>
      </div>
      <div
        style="width: 300px;"
      >
        <div
          style="width: 300px;"
        >
          3
        </div>
      </div>
      <div
        style="width: 300px;"
      >
        <div
          style="width: 300px;"
        >
          4
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Carousel should render correctly with no props 1`] = `
<div
  class="container"
  data-qa="Carousel"
  style="--position: 0px;"
>
  <div
    class="header container direction-horizontal justify-space-between"
    data-qa="Space"
  >
    <div
      class="title"
    />
    <div
      class="arrows container direction-horizontal align-end"
      data-qa="Space"
      style="column-gap: 12px; row-gap: 12px;"
    >
      <span
        class="arrow leftArrow disabled container size-16 asContainer shape-round"
        data-qa="Icon"
      >
        <svg
          data-qa="ArrowLeft"
          fill="none"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m13 4-6 6 6 6"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </span>
      <span
        class="arrow rightArrow container size-16 asContainer shape-round"
        data-qa="Icon"
      >
        <svg
          data-qa="ArrowRight"
          fill="none"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m7 16 6-6-6-6"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    class="slides rightGradient"
  >
    <div
      class="wrapper"
    />
  </div>
</div>
`;

exports[`Carousel should render correctly with slides 1`] = `
<div
  class="container"
  data-qa="Carousel"
  style="--position: 0px;"
>
  <div
    class="header container direction-horizontal justify-space-between"
    data-qa="Space"
  >
    <div
      class="title"
    />
    <div
      class="arrows container direction-horizontal align-end"
      data-qa="Space"
      style="column-gap: 12px; row-gap: 12px;"
    >
      <span
        class="arrow leftArrow disabled container size-16 asContainer shape-round"
        data-qa="Icon"
      >
        <svg
          data-qa="ArrowLeft"
          fill="none"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m13 4-6 6 6 6"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </span>
      <span
        class="arrow rightArrow container size-16 asContainer shape-round"
        data-qa="Icon"
      >
        <svg
          data-qa="ArrowRight"
          fill="none"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m7 16 6-6-6-6"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    class="slides rightGradient"
  >
    <div
      class="wrapper"
    >
      <div>
        <div>
          1
        </div>
      </div>
      <div>
        <div>
          2
        </div>
      </div>
      <div>
        <div>
          3
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Carousel should render correctly with slides and gradient 1`] = `
<div
  class="container"
  data-qa="Carousel"
  style="--position: 0px;"
>
  <div
    class="header container direction-horizontal justify-space-between"
    data-qa="Space"
  >
    <div
      class="title"
    />
    <div
      class="arrows container direction-horizontal align-end"
      data-qa="Space"
      style="column-gap: 12px; row-gap: 12px;"
    >
      <span
        class="arrow leftArrow disabled container size-16 asContainer shape-round"
        data-qa="Icon"
      >
        <svg
          data-qa="ArrowLeft"
          fill="none"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m13 4-6 6 6 6"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </span>
      <span
        class="arrow rightArrow container size-16 asContainer shape-round"
        data-qa="Icon"
      >
        <svg
          data-qa="ArrowRight"
          fill="none"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m7 16 6-6-6-6"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    class="slides gradient rightGradient"
  >
    <div
      class="wrapper"
    >
      <div
        style="width: 300px;"
      >
        <div
          style="width: 300px;"
        >
          1
        </div>
      </div>
      <div
        style="width: 300px;"
      >
        <div
          style="width: 300px;"
        >
          2
        </div>
      </div>
      <div
        style="width: 300px;"
      >
        <div
          style="width: 300px;"
        >
          3
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Carousel should render with title 1`] = `
<div
  class="container"
  data-qa="Carousel"
  style="--position: 0px;"
>
  <div
    class="header container direction-horizontal justify-space-between"
    data-qa="Space"
  >
    <div
      class="title"
    >
      Title
    </div>
    <div
      class="arrows container direction-horizontal align-end"
      data-qa="Space"
      style="column-gap: 12px; row-gap: 12px;"
    >
      <span
        class="arrow leftArrow disabled container size-16 asContainer shape-round"
        data-qa="Icon"
      >
        <svg
          data-qa="ArrowLeft"
          fill="none"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m13 4-6 6 6 6"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </span>
      <span
        class="arrow rightArrow container size-16 asContainer shape-round"
        data-qa="Icon"
      >
        <svg
          data-qa="ArrowRight"
          fill="none"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m7 16 6-6-6-6"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    class="slides rightGradient"
  >
    <div
      class="wrapper"
    />
  </div>
</div>
`;
